### DISCUSSION TITLE
esp32 uart pin mapping

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12203] [Hardware & Peripherals]
I'm a bit confused about pin mapping. On a board like https://www.wemos.cc/en/latest/_static/files/sch_s2_mini_v1.0.0.pdf there are no Tx/Rx labels on the pcb artwork & the user accessible uart (uart1 in the case of the s2) can be mapped to any pair of pins with something like
```
uart=UART(1, 9600, tx=12, rx=13)
```
However other boards like https://www.wemos.cc/en/latest/_static/files/sch_c3_pico_v1.0.0.pdf do have Tx/Rx labels on the pcb artwork. Am I correct in thinking the user accessible uart on these boards is fixed on those pins & can not be moved elsewhere with the uart= command?

**Comment by karfas** [comment]
> Am I correct in thinking the user accessible uart on these boards is fixed on those pins & can not be moved elsewhere with the uart= command?

No. The ESP32 can map most of it's functions to arbitrary GPIOs (see the datasheet). Nothing *printed* on a PCB will be able to take this function away from the processor.
Of course it's way more *convenient* to use the GPIOs as designed by the board manufacturer.

**Reply by jimmo** [reply]
That's right

@kjm1102 To elaborate slightly further, there are three things relevant here:
 - ESP32 can pretty much make any pin do any function (different to the "alternate function" mappings that most other MCUs in this class implement).
 - You can set a default pin for a function. The ESP32 has a bunch of defaults, but a MicroPython board definition may override this, and would typically do so to match the silkscreen.
 - As of about a fortnight ago, MicroPython on the ESP32 supports board naming of pins (which other ports have supported for a long time). i.e. you can use `machine.Pin("FOO")` or `machine.Pin.board.FOO` to access pins by their "board" name (typically what's written on the silkscreen).

The last point isn't specifically relevant to your case, except that if your board didn't do (2), but did to (3) then you could now write `machine.UART(..., tx=machine.Pin.board.TX, rx=machine.Pin.board.RX)`

**Reply by kjm1102** [reply]
I guess some might cope with another layer of abstraction to add to the pin number, gpio number we already have. For the simple minded, like me, I can think of more pressing issues in upython that I would be fixing if I had the skill set.

**Reply by robert-hh** [reply]
>  to add to the pin number, gpio number

Since that same time one can as well use a number or a string to denote a Pin, whenever a Pin has to be specified, like `machine.Pin(4)`. At ESP32, this is usually the GPIO number. When By context it has to be a Pin, one can omit the Pin() call. So the lines below are equivalent:

```
from machine import UART, Pin
uart = UART(1, rx=Pin(4), tx=Pin(5))
uart = UART(1, rx=4, tx=5)
# assuming that pins.csv there are lines assigning RX and TX to GPIO numbers
uart = UART(1, rx=Pin("RX"), tx=Pin("TX"))
uart = UART(1, rx="RX", tx="TX")
uart = UART(1, rx=Pin.board.RX, tx=Pin.board.TX)
```

Similar for all other uses of a Pin.
